### ISSUE TITLE
Is the latest GraalVM EE 22.3.2 supposed to be slower than OpenJDK when executing the included Java program? The native image is also much slower, even with profiling information.

### ISSUE THREAD

**Issue by saoj** [oracle/graal#6564]
### Context

I'm doing a research project where I'm comparing `AOT vs JIT`, `Java vs C++` and `GraalVM vs OpenJDK`. _I am a big fan of Graal._ My intention is that the _step-by-step_ examples below can somehow help you guys make GraalVM even better and faster.

### Summary
**Iterations:** 10 million passes discarding first 1 million for warmup
**Note:** raw minimum latency in nanoseconds (for average see results further below)

Oracle JDK 17 => **806**
OpenJDK 17 => **811**
GraalVM EE 22.3.2 with OpenJDK 17 => **1196** (_not sure why it is worse... different JIT for GraalVM?_)
GraalVM EE 22.3.2 with OpenJDK 17 with `native-image` => **2215** (_native code_)
GraalVM EE 22.3.2 with OpenJDK 17 with `native-image` and profiling through `--pgo-instrument` => **1667** (_native code_)

#### Linux Environment
```
$ uname -a
Linux hivelocity 4.15.0-20-generic #21-Ubuntu SMP Tue Apr 24 06:16:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ cat /etc/issue | head -n 1
Ubuntu 18.04.6 LTS \n \l

$ gcc --version | head -n 1
gcc (Ubuntu 11.1.0-1ubuntu1~18.04.1) 11.1.0

$ cat /proc/cpuinfo | grep "model name" | head -n 1 | awk -F ": " '{print $NF}'
Intel(R) Xeon(R) E-2288G CPU @ 3.70GHz

$ arch
x86_64
```
#### JVMs
```
java version "17.0.1" 2021-10-19 LTS
Java(TM) SE Runtime Environment (build 17.0.1+12-LTS-39)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.1+12-LTS-39, mixed mode, sharing)

openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment (build 17.0.1+12-39)
OpenJDK 64-Bit Server VM (build 17.0.1+12-39, mixed mode, sharing)

java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15)
Java HotSpot(TM) 64-Bit Server VM GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15, mixed mode, sharing)

$ native-image --version
GraalVM 22.3.2 Java 17 EE (Java Version 17.0.7+8-LTS-jvmci-22.3-b15)
```
### Deterministic BubbleSort Java program
**Note:** This Java program is coded in a way to _prevent any kind of code removal by the Java compiler and JIT_, in other words, any important code that needs to be measured by the benchmark is being kept by the Java compiler and JIT during their optimizations.
```java
public class TimeBubbleSort {

    private static void swapping(int[] array, int x, int y) {
        int temp = array[x];
        array[x] = array[y];
        array[y] = temp;
    }

    private static void bubbleSort(int[] array, int size) {
        for(int i = 0; i < size; i++) {
            boolean swaps = false; // flag to detect any swap is there or not
            for(int j = 0; j < size - i - 1; j++) {
                if (array[j] > array[j + 1]) { // when the current item is bigger than next
                    swapping(array, j, j + 1);
                    swaps = true;
                }
            }
            if (swaps == false) break; // No swap in this pass, so array is sorted
        }
    }

    private final static void doSomething(int[] array, int size) {

        // always reset the array to the same worst scenario before sorting...
        for(int z = 0; z < size; z++) {
            array[z] = size - z;
        }

        bubbleSort(array, size);
    }

    public static void main(String[] args) {

        int iterations = Integer.parseInt(args[0]);
        int warmup = Integer.parseInt(args[1]);
        int arraySize = Integer.parseInt(args[2]);

        long x = 0;
        long totalTime = 0;
        long minTime = Long.MAX_VALUE;

        int[] array = new int[arraySize];

        for(int i = 0; i < iterations; i++) {

            long start = System.nanoTime();
            doSomething(array, arraySize);
            long end = System.nanoTime();

            // this code is here just to calculate a final value so that 
            // the JIT/JVM does not try to remove any previous code
            // it is also important so that we can check that we are always 
            // getting the same result for deterministic behavior
            for(int j = 0; j < arraySize; j++) {
                x += array[j];
            }

            // Calculate the latency here: end - start
            int res = (int) (end - start);
            if (res <= 0) res = 1;
            if (i >= warmup) {
                totalTime += res;
                minTime = Math.min(minTime, res);
            }
        }

        int count = iterations - warmup;
        double avg = totalTime / count;
        StringBuilder sb = new StringBuilder();
        sb.append("Value computed: ").append(x).append("\n");
        sb.append("Iterations: ").append(count).append(" | Avg Time: ").append(avg);
        sb.append(" | Min Time: ").append(minTime);
        System.out.println(sb.toString() + "\n");
    }
}
```
## Step-by-Step with Results
**Note** that proper _**JVM warmup**_ of 1 million passes is being performed by the benchmarks before timing starts.
#### Oracle Java 17 (minTime=806)
```
$ java -version
java version "17.0.1" 2021-10-19 LTS
Java(TM) SE Runtime Environment (build 17.0.1+12-LTS-39)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.1+12-LTS-39, mixed mode, sharing)

$ javac TimeBubbleSort.java

$ java TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 831.0 | Min Time: 806
```
#### OpenJDK 17 (minTime=811)
```
$ java -version
openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment (build 17.0.1+12-39)
OpenJDK 64-Bit Server VM (build 17.0.1+12-39, mixed mode, sharing)

$ javac TimeBubbleSort.java

$ java TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 831.0 | Min Time: 811
```
#### GraalVM Enterprise Edition 22.3.2 with OpenJDK 17 (minTime=1196)
```
$ java -version
java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15)
Java HotSpot(TM) 64-Bit Server VM GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15, mixed mode, sharing)

$ javac TimeBubbleSort.java

$ java TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 1412.0 | Min Time: 1196
```
#### GraalVM Enterprise Edition 22.3.2 with `native-image` (minTime=2215)
**Note:** I've tried all the options below, but the results were exactly the same:
```
native-image -O2 TimeBubbleSort -o TimeBubbleSort
native-image -O3 TimeBubbleSort -o TimeBubbleSort
native-image TimeBubbleSort -o TimeBubbleSort --native-compiler-options="-O2"
native-image TimeBubbleSort -o TimeBubbleSort --native-compiler-options="-O3"
```
```
$ java -version
java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15)
Java HotSpot(TM) 64-Bit Server VM GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15, mixed mode, sharing)

$ javac TimeBubbleSort.java

$ native-image TimeBubbleSort -o TimeBubbleSort
========================================================================================================================
GraalVM Native Image: Generating 'TimeBubbleSort' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                    (5.0s @ 0.22GB)
 Version info: 'GraalVM 22.3.2 Java 17 EE'
 Java version info: '17.0.7+8-LTS-jvmci-22.3-b15'
 C compiler: gcc (linux, x86_64, 11.1.0)
 Garbage collector: Serial GC
[2/7] Performing analysis...  [*****]                                                                   (12.1s @ 0.79GB)
   1,842 (61.28%) of  3,006 classes reachable
   1,663 (46.54%) of  3,573 fields reachable
   7,246 (34.50%) of 21,000 methods reachable
      21 classes,     0 fields, and   253 methods registered for reflection
      49 classes,    32 fields, and    48 methods registered for JNI access
       4 native libraries: dl, pthread, rt, z
[3/7] Building universe...                                                                               (1.7s @ 1.04GB)
[4/7] Parsing methods...      [*]                                                                        (1.5s @ 0.44GB)
[5/7] Inlining methods...     [***]                                                                      (1.1s @ 0.64GB)
[6/7] Compiling methods...    [*****]                                                                   (26.0s @ 2.23GB)
[7/7] Creating image...                                                                                  (1.0s @ 2.43GB)
   2.28MB (38.27%) for code area:     3,320 compilation units
   3.38MB (56.71%) for image heap:   43,505 objects and 1 resources
 306.15KB ( 5.02%) for other data
   5.95MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
 289.91KB java.lang                                          445.31KB byte[] for code metadata
 184.27KB java.util                                          351.85KB byte[] for java.lang.String
 147.94KB com.oracle.svm.core.code                           292.57KB java.lang.String
 141.71KB com.oracle.svm.core.genscavenge                    270.78KB java.lang.Class
 113.01KB java.util.concurrent                               232.74KB byte[] for general heap data
  90.70KB java.math                                          111.71KB char[]
  89.20KB com.oracle.svm.core.jni.functions                  104.53KB java.util.HashMap$Node
  74.45KB java.lang.invoke                                    78.67KB java.util.HashMap$Node[]
  68.09KB com.oracle.svm.core                                 71.95KB com.oracle.svm.core.hub.DynamicHubCompanion
  66.21KB java.io                                             66.41KB java.lang.Object[]
   1.02MB for 95 more packages                               425.36KB for 469 more object types
------------------------------------------------------------------------------------------------------------------------
                        0.7s (1.4% of total time) in 17 GCs | Peak RSS: 3.09GB | CPU load: 1.98
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 /home/soliveira/graal/TimeBubbleSort (executable)
 /home/soliveira/graal/TimeBubbleSort.build_artifacts.txt (txt)
========================================================================================================================
Finished generating 'TimeBubbleSort' in 49.6s.

$ ./TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 2299.0 | Min Time: 2215
```
#### GraalVM Enterprise Edition 22.3.2 with `native-image --pgo-instrument` (minTime=1667)
```
$ java -version
java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15)
Java HotSpot(TM) 64-Bit Server VM GraalVM EE 22.3.2 (build 17.0.7+8-LTS-jvmci-22.3-b15, mixed mode, sharing)

$ javac TimeBubbleSort.java

$ native-image --pgo-instrument TimeBubbleSort -o TimeBubbleSort
========================================================================================================================
GraalVM Native Image: Generating 'TimeBubbleSort' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                    (6.9s @ 0.23GB)
 Version info: 'GraalVM 22.3.2 Java 17 EE'
 Java version info: '17.0.7+8-LTS-jvmci-22.3-b15'
 C compiler: gcc (linux, x86_64, 11.1.0)
 Garbage collector: Serial GC
[2/7] Performing analysis...  [******]                                                                  (34.8s @ 1.01GB)
   3,634 (75.44%) of  4,817 classes reachable
   4,403 (52.20%) of  8,435 fields reachable
  17,279 (47.23%) of 36,582 methods reachable
     160 classes,     0 fields, and   545 methods registered for reflection
      58 classes,    58 fields, and    52 methods registered for JNI access
       4 native libraries: dl, pthread, rt, z
[3/7] Building universe...                                                                               (3.8s @ 1.60GB)
[4/7] Parsing methods...      [***]                                                                      (5.3s @ 1.21GB)
[5/7] Inlining methods...     [**]                                                                       (0.7s @ 1.45GB)
[6/7] Compiling methods...    [[6/7] Compiling methods...    [*********]                                                               (94.3s @ 2.25GB)
[7/7] Creating image...                                                                                  (9.8s @ 1.11GB)
  10.28MB (22.76%) for code area:     9,306 compilation units
  32.13MB (71.15%) for image heap:  965,529 objects and 5 resources
   2.75MB ( 6.09%) for other data
  45.16MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
   1.20MB java.util                                            4.90MB java.lang.String
 675.77KB java.lang                                            4.84MB byte[] for java.lang.String
 603.55KB com.sun.crypto.provider                              4.54MB java.util.concurrent.ConcurrentHashMap$Node
 528.21KB com.oracle.svm.core.code                             3.79MB int[]
 511.35KB java.text                                            3.56MB byte[] for general heap data
 447.28KB java.util.concurrent                                 1.77MB byte[] for code metadata
 318.09KB java.util.regex                                      1.36MB c.o.s.e.p.d.InstrumentationSup~ionalProfilingPoint
 310.72KB java.math                                            1.14MB java.util.concurrent.ConcurrentHashMap$Node[]
 291.98KB sun.security.provider                                1.08MB c.o.s.e.p.d.InstrumentationSup~tnessProfilingPoint
 275.26KB java.io                                            857.39KB com.oracle.svm.enterprise.profiling.elements.c
   5.14MB for 149 more packages                                3.24MB for 894 more object types
------------------------------------------------------------------------------------------------------------------------
                        4.6s (2.9% of total time) in 26 GCs | Peak RSS: 5.15GB | CPU load: 1.00
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 /home/soliveira/graal/TimeBubbleSort (executable)
 /home/soliveira/graal/TimeBubbleSort.build_artifacts.txt (txt)
========================================================================================================================
Finished generating 'TimeBubbleSort' in 2m 38s.

$ ./TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 6622.0 | Min Time: 6494

$ native-image --pgo=default.iprof TimeBubbleSort -o TimeBubbleSort
========================================================================================================================
GraalVM Native Image: Generating 'TimeBubbleSort' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                    (7.1s @ 0.24GB)
 Version info: 'GraalVM 22.3.2 Java 17 EE'
 Java version info: '17.0.7+8-LTS-jvmci-22.3-b15'
 C compiler: gcc (linux, x86_64, 11.1.0)
 Garbage collector: Serial GC
[2/7] Performing analysis...  [*****]                                                                   (16.0s @ 0.82GB)
   1,842 (61.28%) of  3,006 classes reachable
   1,663 (46.54%) of  3,573 fields reachable
   7,246 (34.50%) of 21,000 methods reachable
      21 classes,     0 fields, and   253 methods registered for reflection
      49 classes,    32 fields, and    48 methods registered for JNI access
       4 native libraries: dl, pthread, rt, z
[3/7] Building universe...                                                                               (2.0s @ 1.07GB)
[4/7] Parsing methods...      [**]                                                                       (2.3s @ 0.51GB)
[5/7] Inlining methods...     [***]                                                                      (1.7s @ 0.76GB)
[6/7] Compiling methods...    [[6/7] Compiling methods...    [*****]                                                                   (26.6s @ 2.37GB)
[7/7] Creating image...                                                                                  (1.3s @ 2.58GB)
   1.73MB (33.06%) for code area:     3,853 compilation units
   3.20MB (61.21%) for image heap:   43,594 objects and 1 resources
 307.09KB ( 5.73%) for other data
   5.23MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
 214.71KB java.lang                                          357.18KB byte[] for code metadata
 148.06KB com.oracle.svm.core.code                           352.30KB byte[] for java.lang.String
 142.16KB java.util                                          292.95KB java.lang.String
 124.40KB com.oracle.svm.core.genscavenge                    270.78KB java.lang.Class
  84.03KB java.util.concurrent                               231.86KB byte[] for general heap data
  60.09KB com.oracle.svm.core.jni.functions                  111.71KB char[]
  54.68KB java.lang.invoke                                    79.57KB java.util.HashMap$Node
  53.75KB java.io                                             78.67KB java.util.HashMap$Node[]
  52.74KB java.math                                           71.95KB com.oracle.svm.core.hub.DynamicHubCompanion
  46.59KB com.oracle.svm.core.jdk                             66.41KB java.lang.Object[]
 762.29KB for 96 more packages                               383.50KB for 474 more object types
------------------------------------------------------------------------------------------------------------------------
                        1.0s (1.7% of total time) in 16 GCs | Peak RSS: 2.95GB | CPU load: 1.00
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 /home/soliveira/graal/TimeBubbleSort (executable)
 /home/soliveira/graal/TimeBubbleSort.build_artifacts.txt (txt)
========================================================================================================================
Finished generating 'TimeBubbleSort' in 58.1s.

$ ./TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 1712.0 | Min Time: 1667
```

**Comment by saoj** [comment]
Also, as a curiosity, see below the _**exact same**_ Java program written _line by line_ in C++ together with the benchmark results:
```
$ clang++ --version
Ubuntu clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
#### Results with `-O0`, `-O1`, `-O2` and `-O3`:
```
$ clang++ TimeBubbleSort.cpp -o TimeBubbleSort -O0
$ ./TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 6889 | Min Time: 6796

$ clang++ TimeBubbleSort.cpp -o TimeBubbleSort -O1
$ ./TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 1253 | Min Time: 1197

$ clang++ TimeBubbleSort.cpp -o TimeBubbleSort -O2
$ ./TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 890 | Min Time: 818

$ clang++ TimeBubbleSort.cpp -o TimeBubbleSort -O3
$ ./TimeBubbleSort 10000000 1000000 60
Value computed: 18300000000
Iterations: 9000000 | Avg Time: 846 | Min Time: 829
```
It looks like _**Java is beating C++**_ (811 vs  818) but GraalVM does not beat C++ (2215 vs 818 !!!), not even with profiled native code (1667 vs 818).

```c++
#include <iostream>
#include <limits>
#include <sstream>

using namespace std;

long get_nano_ts(timespec* ts) {
    clock_gettime(CLOCK_MONOTONIC, ts);
    return ts->tv_sec * 1000000000 + ts->tv_nsec;
}

struct mi {
   long value;
};

void swapping(int &a, int &b) {
   int temp;
   temp = a;
   a = b;
   b = temp;
}

void bubbleSort(int *array, int size) {
   for(int i = 0; i < size; i++) {
      bool swaps = false;
      for(int j = 0; j < size - i - 1; j++) {
         if(array[j] > array[j+1]) {
            swapping(array[j], array[j+1]);
            swaps = true;
         }
      }
      if (!swaps) break;
   }
}

void doSomething(int *array, int size) {

    for(int z = 0; z < size; z++) {
        array[z] = size - z;
    }

    bubbleSort(array, size);
}

int main(int argc, char* argv[]) {

    int iterations = stoi(argv[1]);
    int warmup = stoi(argv[2]);
    int arraySize = stoi(argv[3]);

    struct timespec ts;

    long long x = 0;
    long long totalTime = 0;
    int minTime = numeric_limits<int>::max();
    int maxTime = numeric_limits<int>::min();

    int * array = (int*) malloc(arraySize * sizeof(int));

    for(int i = 0; i < iterations; i++) {

        long start = get_nano_ts(&ts);

        doSomething(array, arraySize);

        long end = get_nano_ts(&ts);

        for(int j = 0; j < arraySize; j++) {
            x += array[j];
        }

        int res = end - start;
        if (res <= 0) res = 1;
        if (i >= warmup) {
            totalTime += res;
            minTime = min(minTime, res);
            maxTime = max(maxTime, res);
        }
    }

    int count = iterations - warmup;
    double avg = totalTime / count;
    cout << "Value computed: " << x << endl;
    
    stringstream ss;
    ss << "Iterations: " << count << " | Avg Time: " << avg;
    ss << " | Min Time: " << minTime;
    cout << ss.str() << endl << endl;

    free(array);

    return 0;
}
```

**Comment by davleopo** [comment]
@saoj  thanks for the test program. We will have a look. 
Out of curiosity - why did you not write & design the benchmark as a jmh micro?

**Comment by davleopo** [comment]
Tracked internally as GR-46013.

**Comment by SergejIsbrecht** [comment]
@saoj ,

here is an example how to use JMH with OpenJDK and native-image https://github.com/SergejIsbrecht/Bench_MethodHandle . For the native-image you may need to change the `reflect-config.json` file to your needs.

**Comment by gergo-** [comment]
Thanks for this report @saoj . This benchmark exposes two corner cases in some EE optimization phases. A workaround for one of them is included in the upcoming 23.0 release (mid-June), so you should already see performance improvements there. The second corner case is tied to the fact that you are only testing with an input array that is descending, so the innermost condition will always be true, which causes us to enter a special case. If you test 23.0 with input arrays that are more mixed, you should see much better performance.
I'm working on better fixes for both issues for the September release.

**Comment by saoj** [comment]
> This benchmark exposes two corner cases in some EE optimization phases. A workaround for one of them is included in the upcoming 23.0 release (mid-June), so you should already see performance improvements there.

Awesome!

> The second corner case is tied to the fact that you are only testing with an input array that is descending, so the innermost condition will always be true, which causes us to enter a special case. If you test 23.0 with input arrays that are more mixed, you should see much better performance.

Sounds good! I'll choose a different array configuration (i.e. more mixed) to test when 23.0 is out. Notice that I cannot randomize the array because this would of course randomize the performance results. But I can choose a more mixed array and re-use this same array on every iterator.

**Comment by 8Keep** [comment]
Can you also run these benchmarks with the community edition?
